lib/Dialect/TritonGPU/Transforms/OptimizeThreadLocality.cpp

```diff
@@ -0,0 +1,278 @@
+      auto accumOperand = user->getOperand(accumOperandNumber);
+      assert(accumOperand.isa<BlockArgument>());
+      auto blockArg = accumOperand.dyn_cast<BlockArgument>();
+      auto blockArgNum = blockArg.getArgNumber();
+      auto forOp = dyn_cast<scf::ForOp>(blockArg.getOwner()->getParentOp());
+      // get oldAccum
+      auto oldAccum =
+          forOp.getInitArgs()[blockArgNum - forOp.getNumInductionVars()];
+      // get old loop user
+      Value loopResult =
+          forOp.getResult(blockArgNum - forOp.getNumInductionVars());
+      assert(loopResult.hasOneUse());
+      OpOperand &loopUse = *(loopResult.getUses().begin());
+      Operation *loopUser = loopUse.getOwner();
+      // get old loop yield
+      auto oldYield = cast<scf::YieldOp>(forOp.getBody()->getTerminator());
+      // create newAccum initialization
+      auto newAccum =
+          createAccum(builder, oldAccum, viewOpTensorShape, slice2d);
+      // create new loop by copying the old for op signature and appending
+      // newAccum to the block arguments
+      auto newLoop = replaceForOpWithNewSignature(
+          builder, forOp, ValueRange{newAccum->getResult(0)});
+      // create thread local reduction (also adds viewOps)
+      auto newReduce = createReduce(builder, reduce, viewOpTensorType);
+
+      // create new accum update
+      auto newUpdate = createUpdate(builder, newLoop, newReduce, user);
+      // create new yield
+      auto newYield = createYield(builder, newLoop, oldYield,
+                                  newUpdate->getResult(0), blockArgNum);
+      // create post loop reduction on the original reduce axis
+      auto newReduce2 = createPostLoopReduce(builder, newLoop, reduce);
+
+      // Replace loop user with new reduce
+      replaceLoopUser(builder, loopUse, newReduce2);
+
+      oldYield.erase();
+      forOp.erase();
+    }
+  };
+
+private:
+  void replaceLoopUser(OpBuilder &builder, OpOperand &loopUse,
+                       Operation *newReduce2) const {
+    Operation *loopUser = loopUse.getOwner();
+    if (auto cvtLayout = dyn_cast<triton::gpu::ConvertLayoutOp>(loopUser)) {
+      builder.setInsertionPointAfter(loopUser);
+      IRMapping mapping;
+      mapping.map(loopUser->getOperands()[0], newReduce2->getResult(0));
+      auto newCvt = builder.clone(*loopUser, mapping);
+      loopUser->replaceAllUsesWith(newCvt);
+      loopUser->erase();
+    } else {
+      builder.setInsertionPoint(loopUser);
+      auto operandIdx = loopUse.getOperandNumber();
+      Value operand = loopUser->getOperand(operandIdx);
+      Type destType = operand.getType();
+      auto newCvt = builder.create<triton::gpu::ConvertLayoutOp>(
+          loopUser->getLoc(), destType, newReduce2->getResult(0));
+      IRMapping mapping;
+      mapping.map(operand, newCvt->getResult(0));
+      auto newUser = builder.clone(*loopUser, mapping);
+      loopUser->replaceAllUsesWith(newUser);
+      loopUser->erase();
+    }
```

Comment:
I don't understand why we need to clone the loop user. It seems fragile and non intuitive. I was expecting that after the new loop we would create a reduction to get to the same shape as the original result and if needed we can create a new convert. We should then just let the convert get optimized away.

Reply:
Make sense, I wasn't sure if the convert would get optimized away. Thanks, will fix

Reply:
I'm assuming `add_tritongpu_remove_layout_conversions_pass` would combine the two `convert_layout` ops into one. Currently it does not run after the new thread locality pass, should I run it after it as well ?

Reply:
Canonicalization pass should be enough